lib/Target/CBackend/CBackend.cpp

```diff
@@ -906,8 +906,14 @@ bool CWriter::printConstantString(Constant *C, enum OperandContext Context) {
 
   StringRef Bytes = CDS->getAsString();
 
-  // Do not include the last character, which we know is null
-  for (unsigned i = 0, e = Bytes.size() - 1; i < e; ++i) {
+  unsigned length = Bytes.size();
+  // We can skip the last character only if it is an implied null.
+  // Beware: C does not force character (i.e. (u)int8_t here) arrays to have a
+  // null terminator, but if the length is not specified it will imply one!
+  if (length >= 1 && Bytes[length - 1] == '\0')
+    length--;
```

Comment:
One consequence of this code is that you can't tell at-a-glance (without counting characters) whether a string is null-terminated or not. It's not very idiomatic, but maybe we should include the `\0` at the end of the string literal so you can distinguish these. 

I think there's a risk of triggering some sort of compiler warning, though.